[Validator] Deprecate `ValidatorBuilder::enableAnnotationMapping()` and `ValidatorBuilder::disableAnnotationMapping()`

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Part of https://github.com/symfony/symfony/issues/51381
| License       | MIT
| Doc PR        | -

Required by https://github.com/symfony/symfony/pull/51425

alexandre-daubois committed:
[Validator] Deprecate `ValidatorBuilder::enableAnnotationMapping()` and `ValidatorBuilder::disableAnnotationMapping()`

stof commented on UPGRADE-6.4.md:
```diff
@@ -156,4 +156,6 @@ Validator
  * Deprecate passing an annotation reader to the constructor signature of `AnnotationLoader`
  * Deprecate `ValidatorBuilder::setDoctrineAnnotationReader()`
- * Deprecate `ValidatorBuilder::addDefaultDoctrineAnnotationReader()`
+ * Deprecate `ValidatorBuilder::addDefaultDoctrineAnnotationReader()
```
missing backtick

stof commented on src/Symfony/Component/Validator/Tests/ValidatorBuilderTest.php:
```diff
@@ -116,9 +116,27 @@ public function testEnableAnnotationMappingWithCustomDoctrineAnnotationReader()
+    public function testExpectDeprecationWhenDisablingAnnotationMapping()
+    {
+        $this->expectDeprecation('Since symfony/validator 6.4: Method "Symfony\Component\Validator\ValidatorBuilder::disableAnnotationMapping()" is deprecated, use "disableAttributeMapping()" instead.');
+        $this->builder->disableAnnotationMapping();
```
we should not loose the assertion about the fluent API that we had before

stof commented on src/Symfony/Component/Validator/Tests/ValidatorBuilderTest.php:
```diff
@@ -81,7 +81,7 @@ public function testAddMethodMappings()
     public function testEnableAnnotationMappingWithDefaultDoctrineAnnotationReader()
     {
-        $this->assertSame($this->builder, $this->builder->enableAnnotationMapping());
+        $this->assertSame($this->builder, $this->builder->enableAttributeMapping());
```
this test is really about annotation mapping. So I would rather expect both deprecations in it than changing method

stof commented on src/Symfony/Component/Validator/Tests/ValidatorBuilderTest.php:
```diff
@@ -102,7 +102,7 @@ public function testEnableAnnotationMappingWithCustomDoctrineAnnotationReader()
         $reader = $this->createMock(Reader::class);
 
-        $this->assertSame($this->builder, $this->builder->enableAnnotationMapping());
+        $this->assertSame($this->builder, $this->builder->enableAttributeMapping());
```
same here

stof commented on src/Symfony/Component/Validator/ValidatorBuilder.php:
```diff
@@ -187,31 +187,56 @@ public function addMethodMappings(array $methodNames): static
+    public function enableAttributeMapping(): static
     {
         if (null !== $this->metadataFactory) {
             throw new ValidatorException('You cannot enable annotation mapping after setting a custom metadata factory. Configure your metadata factory instead.');
```
this exception message should be updated

stof commented on src/Symfony/Component/Validator/ValidatorBuilder.php:
```diff
@@ -344,7 +369,7 @@ public function getLoaders(): array
 
-        if ($this->enableAnnotationMapping) {
+        if ($this->enableAttributeMapping) {
             $loaders[] = new AnnotationLoader($this->annotationReader);
```
if you don't rename the loader class as well, this work is incomplete

xabbuh reviewed (changes requested):
We also need to update the `FrameworkExtension`.

alexandre-daubois commented:
Alright, I thought it was the way to deal with the CI that misses some not-yet-merged methods. Closing in favor of https://github.com/symfony/symfony/pull/51425, I addressed your comments there :slightly_smiling_face:
